tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+//
+#include <range_algorithm_support.hpp>
+
+constexpr void smoke_test() {
+    using ranges::find_end, ranges::iterator_t, ranges::subrange, std::array, std::same_as;
+    using P = std::pair<int, int>;
+
+    // Validate dangling story
+    STATIC_ASSERT(same_as<decltype(find_end(borrowed<false>{}, array<int, 42>{})), ranges::dangling>);
+    STATIC_ASSERT(same_as<decltype(find_end(borrowed<true>{}, array<int, 42>{})), subrange<int*>>);
+
+    const array pairs = {P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{0, 42}};
+
+    const auto pred = [](const int x, const int y) { return x == y + 1; };
+
+    {
+        // Validate range overload [found case]
+        const auto result = find_end(pairs, array{-1, 0}, pred, get_first);
+        STATIC_ASSERT(same_as<decltype(result), const subrange<iterator_t<decltype(pairs)>>>);
+        assert(result.size() == 2);
+        assert(result.begin() == pairs.begin() + 4);
+        assert(result.end() == pairs.begin() + 6);
+    }
+    {
+        // Validate iterator + sentinel overload [found case]
+        const array needle = {-1, 0};

```

Comment:
This confuses me, you are not declaring the temporäre needle in the other cases and I have to figure out why.
 Could you just always declare needle. I believe in tests readability wins. (Even if we do not need it because of range interface)